### PR TITLE
Upgrades TypeScript to 4.1, simplifies jest config

### LINKED ISSUE #1081: CVE-2021-3803 (High) detected in nth-check-1.0.2.tgz
## CVE-2021-3803 - High Severity Vulnerability
<details><summary> Vulnerable Library - <b>nth-check-1.0.2.tgz</b></p></summary>

<p>performant nth-check parser & compiler</p>
<p>Library home page: <a href="https://registry.npmjs.org/nth-check/-/nth-check-1.0.2.tgz">https://registry.npmjs.org/nth-check/-/nth-check-1.0.2.tgz</a></p>
<p>

Dependency Hierarchy:
  - cheerio-0.22.0.tgz (Root Library)
    - css-select-1.2.0.tgz
      - :x: **nth-check-1.0.2.tgz** (Vulnerable Library)
<p>Found in HEAD commit: <a href="https://github.com/opensearch-project/OpenSearch-Dashboards/commit/4fd064970b66ce555f48c22dfab6ed965d0e260a">4fd064970b66ce555f48c22dfab6ed965d0e260a</a></p>
<p>Found in base branch: <b>main</b></p>
</p>
</details>
<p></p>
<details><summary> Vulnerability Details</summary>
<p>  
  
nth-check is vulnerable to Inefficient Regular Expression Complexity

<p>Publish Date: 2021-09-17
<p>URL: <a href=https://vuln.whitesourcesoftware.com/vulnerability/CVE-2021-3803>CVE-2021-3803</a></p>
</p>
</details>
<p></p>
<details><summary> CVSS 3 Score Details (<b>7.5</b>)</summary>
<p>

Base Score Metrics:
- Exploitability Metrics:
  - Attack Vector: Network
  - Attack Complexity: Low
  - Privileges Required: None
  - User Interaction: None
  - Scope: Unchanged
- Impact Metrics:
  - Confidentiality Impact: None
  - Integrity Impact: None
  - Availability Impact: High
</p>
For more information on CVSS3 Scores, click <a href="https://www.first.org/cvss/calculator/3.0">here</a>.
</p>
</details>
<p></p>
<details><summary> Suggested Fix</summary>
<p>

<p>Type: Upgrade version</p>
<p>Origin: <a href="https://github.com/fb55/nth-check/compare/v2.0.0...v2.0.1">https://github.com/fb55/nth-check/compare/v2.0.0...v2.0.1</a></p>
<p>Release Date: 2021-09-17</p>
<p>Fix Resolution (nth-check): 2.0.1</p>
<p>Direct dependency fix Resolution (cheerio): 1.0.0-rc.5</p>

</p>
</details>
<p></p>

### PR DESCRIPTION
**Pull request by tmarkley**
### Description

Most of the PR is just adding parameters to `Promise`s and changing references to the updated `@osd/test/jest` directory.

* Addressing the `nth-check` CVE required bumping `css-select`, which is a dependency of `cheerio`. Bumping `cheerio` requires upgrading from TypeScript 4.0 to 4.1.
* TypeScript 4.1 introduces a set of [breaking changes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#breaking-changes). The main changes that impact Dashboards is that `resolve`'s parameters are no longer optional in `Promise`s, and that potentially undefined indexes must use the `!` non-null assertion operator.
* The upgrades to TypeScript and `cheerio` triggered some jest errors which prompted the upgrade to the `enzyme` dependencies.
* Merges files under `/src/test_utils` and `/src/dev/jest` into the `@osd/test` package to simplify.
* Fixes the naming of the `@osd/eslint-config-opensearch-dashboards` package.
 
### Issues Resolved

Resolves #1081

Partially addresses #1187 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
    - [x] `yarn test:jest`
    - [x] `yarn test:jest_integration`
    - [x] `yarn test:ftr`
- [x] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by tmarkley** [comment]
Currently trying to work through bootstrap errors related to `istanbul-lib-coverage` types. It may be easier to resolve this problem after #1300 and #1301 are merged.

**Comment by tmarkley** [comment]
> > Merges files under /src/test_utils and /src/dev/jest into the @osd/test package to simplify.
> 
> Would this reduce the delta changed in this PR but resolve the CVE if the merging happened in a separate PR?

Yeah, but I can't get the TypeScript upgrade to bootstrap even without those jest changes. There's a conflict with the jest dependencies so I'm going to wait for #1301 and then try again.

**Comment by tmarkley** [comment]
> Currently trying to fix the bootstrap errors related to istanbul-lib-coverage types. It may be easier to resolve this problem after https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1300 and https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1301 are merged.

This has been resolved after merging those PRs and rebasing.

File: src/cli/serve/integration_tests/reload_logging_config.test.ts [hunk 1]
```diff
@@ -135,7 +135,7 @@ describe('Server logging configuration', function () {
           '--verbose',
         ]);
 
-        const message$ = Rx.fromEvent(child.stdout, 'data').pipe(
+        const message$ = Rx.fromEvent(child.stdout!, 'data').pipe(
```
Comment (ashwin-pc) [inline src/cli/serve/integration_tests/reload_logging_config.test.ts:138]: Nit: This is [nullable](https://nodejs.org/api/child_process.html#subprocessstdout). Given that its a test it should be okay though.

**Review by ashwin-pc — APPROVED** [review]
LGTM

**Comment by kavilla** [comment]
Did we want to put this PR in draft?

**Comment by tmarkley** [comment]
> Did we want to put this PR in draft?

Good question. I'm still actively trying to fix the functional test error, but I guess I can put it in draft until that's fixed.

**Comment by kavilla** [comment]
Looks like the CVE was addressed in this [PR](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/1422). For that reason and the conflicts, I will remove the `2.0.0`.

**Comment by kavilla** [comment]
Verify impact to plugins.

**Comment by seanneumann** [comment]
Closing. We are not doing a TypeScript upgrade now as the refactor is too much for the benefit. We can revisit this later.

**Pull request closed without merging**